File: apple/HavenApp/SelfSync.swift
import Foundation

// Multi-device live sync (roadmap D16, Phase 3 — client wiring).
//
// Makes a user's OWN devices converge: each device writes a self-encrypted snapshot of its
// account state to a per-account mailbox slot it owns, and merges its peers' slots. The merge
// is the CRDT in `p2pcore::selfsync` (last-write-wins per key), exposed through the FFI
// (`AccountStateHandle`, `sealAccountState`/`openAccountState`, `selfSyncSlotKey`). The relay
// only ever holds ciphertext sealed with a key only this account's devices can derive.
//
// Scope: PROFILE (name/emoji/bio/link), GLOBAL SETTINGS, CONTACTS, and the BLOCKED LIST.
// Scalar keys (profile/setting) apply via `get(key:)`; set-like state (contacts/blocked)
// reconciles via `entries()`, with local removals propagated as tombstones. Circles stay out
// for now: a device needs each member's public crypto bundle (held in the engine, not the
// Contact struct) to actually participate, so circle sync is a separate engine-state piece.

/// A stable **per-device** id. All of a user's devices share the account seed (same node id),
/// so each physical device needs its own id to own a sync slot and to break LWW ties. Random
/// 32 bytes, generated once, stored device-local in UserDefaults, never synced.
enum SelfSyncDevice {
    private static let key = "haven.selfsync.deviceId"
    static let id: Data = {
        let d = UserDefaults.standard
        if let hex = d.string(forKey: key), let data = Data(havenHex: hex), data.count == 32 { return data }
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, 32, &bytes)
        let data = Data(bytes)
        d.set(data.havenHexString, forKey: key)
        return data
    }()
    static var hex: String { id.havenHexString }
}

@MainActor
final class SelfSyncCoordinator {
    static let shared = SelfSyncCoordinator()
    private init() {}

    private var inFlight = false

    // Circle records are encoded/decoded by the shared FFI `encodeCircleSync`/`decodeCircleSync`
    // so iOS/desktop/Android emit byte-identical bytes (no per-platform JSON/base64 drift).

    /// Last converged state, persisted so we can detect what changed locally (LWW only advances
    /// a key's stamp when its value actually changes — otherwise two devices would ping-pong).
    private var baseURL: URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        return dir.appendingPathComponent("haven-selfsync.bin")
    }

    // MARK: state ↔ CRDT mapping (v1: profile + global settings)

    /// The current local state as namespaced key → value bytes (no stamps). `social` (when
    /// available) contributes circle structure; without it, circles are simply not snapshotted.
    private func currentLocal(social: HavenSocial?) -> [String: Data] {
        var m: [String: Data] = [:]
        let p = ProfileStore.shared
        // Only broadcast NON-EMPTY profile scalars. A fresh/empty device must never stamp a blank value
        // that then wins last-writer-wins and REVERTS a sibling's real profile (absence ≠ authoritative —
        // same principle as the contact/circle tombstone rules). Clearing a field is an explicit action we
        // can model separately; silently blanking a real profile from another device is data loss.
        if !p.displayName.isEmpty { m["profile:name"] = Data(p.displayName.utf8) }
        if !p.emoji.isEmpty { m["profile:emoji"] = Data(p.emoji.utf8) }
        if !p.bio.isEmpty { m["profile:bio"] = Data(p.bio.utf8) }
        if !p.link.isEmpty { m["profile:link"] = Data(p.link.utf8) }
        // Profile photo (small base64 JPEG) — so a freshly-linked device gets the avatar too, not just
        // the name/bio. Was missing, which is why the avatar showed on posts but not on the profile.
        let av = p.avatarBase64
        if !av.isEmpty { m["profile:avatar"] = Data(av.utf8) }
        let s = SettingsStore.shared
        m["setting:saveToPhotos"] = Data([s.saveToPhotos ? 1 : 0])
        m["setting:saveOthersToPhotos"] = Data([s.saveOthersToPhotos ? 1 : 0])
        m["setting:autoOptimize"] = Data([s.autoOptimize ? 1 : 0])
        m["setting:silent"] = Data([s.silent ? 1 : 0])
        m["setting:retentionDays"] = withUnsafeBytes(of: Int32(s.retentionDays).littleEndian) { Data($0) }
        // Roster: contacts (full card) + blocked list.
        for c in ContactsStore.shared.contacts {
            if let data = try? JSONEncoder().encode(c) { m["contact:\(c.idHex)"] = data }
        }
        for hex in ConnectionsStore.shared.blocked { m["blocked:\(hex)"] = Data([1]) }
        // Explicit circle severances — propagate as ADDITIVE, grow-only records (a removal is intentional
        // and must never be undone by a peer's absence; that's why it's NOT a dynamic prefix). Keyed
        // removal:<circleId>|<hex>. This is the safe way to make "remove someone" stick across my devices.
        for key in ConnectionsStore.shared.circleRemovals { m["removal:\(key)"] = Data([1]) }
        // Circles: name + member bundles + relay nodes, so another device can reconstruct each
        // circle and seal to every member. (Additive in v1 — member/circle removal is a follow-up.)
        if let social = social {
            for ci in social.circles() {
                // Shared FFI encoder → byte-identical circle records across iOS/desktop/Android.
                m["circle:\(ci.id)"] = encodeCircleSync(
                    name: ci.name,
                    memberBundles: social.circleMemberBundles(circleId: ci.id),
                    relays: RelayMailboxStore.shared.relays(forCircle: ci.id))
            }
        }
        return m
    }

    /// Namespaces whose keys are dynamic (set-like) — used to detect LOCAL removals so they
    /// propagate as tombstones (unblock, delete contact). Scalar namespaces (profile/setting)
    /// are always present, so they're never spuriously removed.
    private static let dynamicPrefixes = ["contact:", "blocked:", "circle:"]

    /// Write a merged state back into the local stores (only when a value actually differs, to
    /// avoid feedback loops through the stores' didSet broadcasts).
    private func applyLocal(_ h: AccountStateHandle, social: HavenSocial?) {
        let p = ProfileStore.shared
        // Never apply an EMPTY scalar over a non-empty local one (belt-and-suspenders against a blank
        // overwrite — see currentLocal). Non-empty incoming values still win normally.
        if let v = h.get(key: "profile:name"), let s = String(data: v, encoding: .utf8), !s.isEmpty, s != p.displayName {
            HavenLog.sync("apply profile:name '\(p.displayName)' → '\(s)'"); p.displayName = s
        }
        if let v = h.get(key: "profile:emoji"), let s = String(data: v, encoding: .utf8), !s.isEmpty, s != p.emoji { p.emoji = s }
        if let v = h.get(key: "profile:bio"), let s = String(data: v, encoding: .utf8), !s.isEmpty, s != p.bio {
            HavenLog.sync("apply profile:bio (\(p.bio.count)→\(s.count) chars)"); p.bio = s
        }
        if let v = h.get(key: "profile:link"), let s = String(data: v, encoding: .utf8), !s.isEmpty, s != p.link { p.link = s }
        if let v = h.get(key: "profile:avatar"), let b64 = String(data: v, encoding: .utf8), b64 != p.avatarBase64,
           let data = Data(base64Encoded: b64), let img = PlatformImage(data: data) {
            p.setAvatar(img)
        }

        let s = SettingsStore.shared
        if let b = boolValue(h, "setting:saveToPhotos"), b != s.saveToPhotos { s.saveToPhotos = b }
        if let b = boolValue(h, "setting:saveOthersToPhotos"), b != s.saveOthersToPhotos { s.saveOthersToPhotos = b }
        if let b = boolValue(h, "setting:autoOptimize"), b != s.autoOptimize { s.autoOptimize = b }
        if let b = boolValue(h, "setting:silent"), b != s.silent { s.silent = b }
        if let v = h.get(key: "setting:retentionDays"), v.count == 4 {
            let n = Int(Int32(littleEndian: v.withUnsafeBytes { $0.loadUnaligned(as: Int32.self) }))
            if n != s.retentionDays { s.retentionDays = n }
        }

        // Roster reconciliation (set-like — enumerate the converged state via entries()).
        let live = h.entries()

        // Contacts: upsert everything present; drop locals the converged state no longer has
        // (a contact deleted on another device propagated as a tombstone).
        var wantContacts: [String: Contact] = [:]
        for e in live where e.key.hasPrefix("contact:") {
            if let c = try? JSONDecoder().decode(Contact.self, from: e.value) { wantContacts[c.idHex] = c }
        }
        let cs = ContactsStore.shared
        for c in wantContacts.values { cs.syncUpsert(c) }
        // ADDITIVE ONLY — do not remove contacts a peer happens not to have (see the circle note above:
        // a freshly-restored device's empty state must never delete the primary's contacts/posts).

        // Blocked list: reconcile both directions.
        var wantBlocked = Set<String>()
        for e in live where e.key.hasPrefix("blocked:") {
            wantBlocked.insert(String(e.key.dropFirst("blocked:".count)))
        }
        let conn = ConnectionsStore.shared
        for hex in wantBlocked.subtracting(conn.blocked) { conn.block(hex) }
        for hex in conn.blocked.subtracting(wantBlocked) { conn.unblock(hex) }

        // Circle severances synced from another of my devices → apply them here too: record the removal
        // and purge that member from the circle (so removing someone on my phone severs them on my Mac).
        for e in live where e.key.hasPrefix("removal:") {
            let key = String(e.key.dropFirst("removal:".count))   // "<circleId>|<hex>"
            guard let bar = key.firstIndex(of: "|") else { continue }
            let circleId = String(key[key.startIndex..<bar])
            let hex = String(key[key.index(after: bar)...])
            guard !circleId.isEmpty, !hex.isEmpty, !conn.isRemovedFromCircle(hex, circleId: circleId) else { continue }
            conn.removeFromCircle(hex, circleId: circleId)
            social?.removeFromCircle(circleId: circleId, nodeHex: hex)
        }

        // Circles: reconstruct each synced circle — create it, register every member's bundle
        // (so this device can seal to them), and record its relay mailbox(es). Additive in v1.
        if let social = social {
            let existing = social.circles()
            for e in live where e.key.hasPrefix("circle:") {
                let id = String(e.key.dropFirst("circle:".count))
                guard let rec = decodeCircleSync(bytes: e.value) else { continue }
                social.createCircle(id: id, name: rec.name)   // no-op if it already exists
                if let cur = existing.first(where: { $0.id == id }), cur.name != rec.name {
                    social.renameCircle(id: id, name: rec.name)
                }
                // STRICTLY ADDITIVE: register each synced member. We do NOT remove members or leave
                // circles based on a peer's state. Absence-based removal caused catastrophic data loss:
                // a freshly-restored device has an empty engine, so it looked like "every circle/member
                // was removed", which tombstoned + propagated to the primary and wiped its posts. Real
                // circle-leave / member-removal must be driven by an explicit intent, not by absence.
                for bundle in rec.memberBundles {
                    // Don't re-add someone we EXPLICITLY removed from this circle. Additive sync was
                    // re-registering removed members from a peer's roster — which is exactly why "remove
                    // someone" never stuck. An explicit removal wins over a peer still listing them.
                    let hex = bundle.prefix(32).map { String(format: "%02x", $0) }.joined()
                    if conn.isRemovedFromCircle(hex, circleId: id) { continue }
                    _ = try? social.addContactBundle(circleId: id, bundle: bundle)
                }
                for node in rec.relays where !RelayMailboxStore.shared.isForgotten(node) {
                    RelayMailboxStore.shared.add(circleId: id, nodeHex: node)   // skip relays the user forgot
                }
            }
        }
    }

    private func boolValue(_ h: AccountStateHandle, _ key: String) -> Bool? {
        guard let v = h.get(key: key), let first = v.first else { return nil }
        return first == 1
    }

    // MARK: sync

    /// One full sync pass: fold local changes into the base with fresh stamps, merge every peer
    /// slot, apply the converged result locally, persist, and re-publish our own slot. Safe to
    /// call on a timer; coalesces if already running. No-op without an account or any sync
    /// target (a relay or the user's S3 bucket — either works, no relay required).
    /// Returns `true` if the merge brought in changes from another device (so the caller can
    /// persist the engine state + refresh the UI — relevant when circles arrive).
    @discardableResult
    func sync(social: HavenSocial?) async -> Bool {
        guard !inFlight else { return false }
        guard let seed = AccountStore.storedSeed() else { return false }
        let accountHex = AccountStore.currentNodeHex()
        guard !accountHex.isEmpty else { return false }
        let transports = gatherTransports()
        guard !transports.isEmpty else { return false }   // needs a relay OR an S3 bucket
        inFlight = true
        defer { inFlight = false }

        // 1. Base = last converged state (or empty).
        let base: AccountStateHandle
        if let data = try? Data(contentsOf: baseURL), let h = try? AccountStateHandle.fromBytes(bytes: data) {
            base = h
        } else {
            base = AccountStateHandle()
        }

        // 2. Fold in whatever changed locally since last sync (stamp = now, this device).
        let now = UInt64(Date().timeIntervalSince1970 * 1000)
        let device = SelfSyncDevice.id
        let local = currentLocal(social: social)
        for (key, value) in local {
            if base.get(key: key) != value {
                _ = try? base.set(key: key, value: value, ts: now, device: device)
            }
        }
        // Detect local removals in dynamic namespaces (a contact deleted, a peer unblocked) and
        // tombstone them so the removal propagates instead of a peer device re-adding them — but ONLY
        // when the engine isn't freshly-empty (see safeToTombstone: a just-restored device must not
        // tombstone the whole account).
        if safeToTombstone(local: local, base: base) {
            for e in base.entries() where Self.dynamicPrefixes.contains(where: { e.key.hasPrefix($0) }) {
                if local[e.key] == nil {
                    _ = try? base.remove(key: e.key, ts: now, device: device)
                }
            }
        }

        // Snapshot post-fold so we can tell whether the merge below actually brought anything new.
        let preMerge = base.toBytes()

        // 3. Pull + merge every peer slot from every relay/bucket.
        let prefix = "haven/" + selfSyncSlotPrefix(accountNodeHex: accountHex)
        let ownKey = "haven/" + selfSyncSlotKey(accountNodeHex: accountHex, deviceNodeHex: SelfSyncDevice.hex)
        for t in transports {
            let keys = await tList(t, prefix)
            for key in keys where key != ownKey {
                guard let blob = await tFetch(t, key) else { continue }
                if let peer = try? openAccountState(accountSeed: seed, sealed: blob) {
                    base.merge(other: peer)
                }
            }
        }

        let changed = base.toBytes() != preMerge

        // 4. Apply the converged state locally + persist the new base.
        applyLocal(base, social: social)
        try? base.toBytes().write(to: baseURL, options: .atomic)

        // 5. Re-publish our own slot (sealed) to every relay/bucket for redundancy.
        guard let sealed = try? sealAccountState(accountSeed: seed, state: base) else { return changed }
        for t in transports { _ = await tUpload(t, ownKey, sealed) }
        return changed
    }

    // MARK: - Direct (nearby) device-to-device sync — NO relay required
    //
    // Two of the user's own devices on the same Bluetooth/Wi-Fi don't need a relay or S3 to sync:
    // they trade their sealed self-sync slots directly over the nearby mesh. This is the local
    // "handshake" path — `sealedLocalSlot` is what a device offers, `ingestPeerSlot` is how it folds
    // in what it receives.

    private func loadBase() -> AccountStateHandle {
        if let data = try? Data(contentsOf: baseURL), let h = try? AccountStateHandle.fromBytes(bytes: data) { return h }
        return AccountStateHandle()
    }

    /// Fold local changes (with fresh stamps) into `base`, including removal tombstones. Mirrors the
    /// relay `sync()`'s steps 1–2.
    private func foldLocal(into base: AccountStateHandle, social: HavenSocial?) {
        let now = UInt64(Date().timeIntervalSince1970 * 1000)
        let device = SelfSyncDevice.id
        let local = currentLocal(social: social)
        for (key, value) in local where base.get(key: key) != value {
            _ = try? base.set(key: key, value: value, ts: now, device: device)
        }
        if safeToTombstone(local: local, base: base) {
            for e in base.entries() where Self.dynamicPrefixes.contains(where: { e.key.hasPrefix($0) }) {
                if local[e.key] == nil { _ = try? base.remove(key: e.key, ts: now, device: device) }
            }
        }
    }

    /// Whether it's safe to emit removal tombstones for dynamic keys. NOT safe when the engine looks
    /// freshly-empty (no circles) but the base still has circles — that's a just-reset / unready device,
    /// and tombstoning there is precisely what wiped accounts. In that state we only ADD, never remove.
    private func safeToTombstone(local: [String: Data], base: AccountStateHandle) -> Bool {
        let localHasCircle = local.keys.contains { $0.hasPrefix("circle:") }
        let baseHasCircle = base.entries().contains { $0.key.hasPrefix("circle:") }
        return localHasCircle || !baseHasCircle
    }

    /// Erase this device's self-sync base. Used when adopting a DIFFERENT identity (restore/link) or on
    /// factory reset, so a freshly-restored device never diffs its empty engine against a STALE base and
    /// tombstones the account's circles/contacts — the bug that propagated and wiped the primary's posts.
    func reset() { try? FileManager.default.removeItem(at: baseURL) }

    /// This device's sealed self-sync slot, folding in local changes first — the payload to hand a
    /// peer device directly over the nearby mesh. No relay/S3 involved.
    func sealedLocalSlot(social: HavenSocial?) -> Data? {
        guard let seed = AccountStore.storedSeed() else { return nil }
        let base = loadBase()
        foldLocal(into: base, social: social)
        try? base.toBytes().write(to: baseURL, options: .atomic)
        return try? sealAccountState(accountSeed: seed, state: base)
    }

    /// Merge a peer device's sealed slot received over a direct transport, apply + persist. Returns
    /// true if anything new arrived (so the caller can refresh the feed).
    @discardableResult
    func ingestPeerSlot(_ blob: Data, social: HavenSocial?) -> Bool {
        guard let seed = AccountStore.storedSeed(),
              let peer = try? openAccountState(accountSeed: seed, sealed: blob) else { return false }
        let base = loadBase()
        let before = base.toBytes()
        base.merge(other: peer)
        let changed = base.toBytes() != before
        applyLocal(base, social: social)
        try? base.toBytes().write(to: baseURL, options: .atomic)
        return changed
    }

    // MARK: transports (relay + S3 — self-sync works with either, or both)

    private enum Transport { case relay(String); case s3(S3Client) }

    /// Every place this device can read/write its self-sync slots: all configured relays plus
    /// the user's OWN S3 bucket (so sync works with no relay at all — BYO storage is enough).
    private func gatherTransports() -> [Transport] {
        var ts: [Transport] = RelayMailboxStore.shared.allRelays().map { .relay($0) }
        if let s3 = SharedStore.ownerS3() { ts.append(.s3(s3)) }
        return ts
    }

    private func tList(_ t: Transport, _ prefix: String) async -> [String] {
        switch t {
        case .relay(let node):
            guard let c = await RelayClients.client(node) else { return [] }
            return await c.list(prefix: prefix)
        case .s3(let c):
            return (try? await c.listKeys(prefix: prefix)) ?? []
        }
    }

    private func tFetch(_ t: Transport, _ key: String) async -> Data? {
        switch t {
        case .relay(let node):
            guard let c = await RelayClients.client(node) else { return nil }
            return await c.get(key: key)
        case .s3(let c):
            return try? await c.getObject(key: key)
        }
    }

    private func tUpload(_ t: Transport, _ key: String, _ data: Data) async -> Bool {
        switch t {
        case .relay(let node):
            guard let c = await RelayClients.client(node) else { return false }
            do { try await c.put(key: key, data: data); RelayHealth.shared.recordSuccess(node); return true }
            catch { RelayHealth.shared.recordFailure(node); return false }
        case .s3(let c):
            do { try await c.putObject(key: key, data: data); return true } catch { return false }
        }
    }
}

private extension Data {
    var havenHexString: String { map { String(format: "%02x", $0) }.joined() }
    init?(havenHex hex: String) {
        let chars = Array(hex)
        guard chars.count % 2 == 0 else { return nil }
        var bytes = [UInt8](); bytes.reserveCapacity(chars.count / 2)
        var i = 0
        while i < chars.count {
            guard let b = UInt8(String(chars[i...i+1]), radix: 16) else { return nil }
            bytes.append(b); i += 2
        }
        self = Data(bytes)
    }
}

File: desktop/src-tauri/src/selfsync.rs
//! Multi-device **self-sync** for the desktop backend (roadmap D16, Phase 3 — client wiring).
//!
//! Direct port of the iOS `SelfSync.swift` coordinator. Makes a user's OWN devices converge:
//! each device writes a self-encrypted snapshot of its account state to a per-account mailbox
//! slot it owns, and merges its peers' slots. The CRDT itself lives in [`p2pcore::selfsync`]
//! (last-write-wins per key); the relay/S3 only ever holds ciphertext sealed with
//! [`p2pcore::identity::Identity::self_sync_key`] — a key only this account's devices can derive.
//!
//! Scope (must match iOS byte-for-byte cross-platform): PROFILE (name/emoji/bio/link), GLOBAL
//! SETTINGS (retention, host-on-launch), CONTACTS, the BLOCKED list, and CIRCLES (name + member
//! bundles + relay nodes). Profile/setting/blocked/circle values are byte-identical to iOS; the
//! `contact:` value is each platform's own (the Contact structs differ) but stable per-platform.
//!
//! This module holds the **pure** pieces — the device-id, the CRDT key/value mapping, and the
//! deterministic circle encoding. The networked coordinator (`Engine::poll_self_sync`) lives in
//! `engine.rs` because it needs the engine's private transport internals.

use std::collections::BTreeMap;

use crate::store::{Contact, Paths, Prefs};
use haven_ffi::multidevice::{decode_circle_sync, encode_circle_sync};
use haven_ffi::HavenSocial;

/// Load (or first-time generate) this device's stable 32-byte self-sync id. All of a user's
/// devices share the account seed (same node id), so each physical device needs its own id to
/// own a sync slot and to break LWW ties. Random, generated once, stored device-local in the
/// Paths dir (`selfsync-device.bin`), and NEVER synced.
pub fn device_id(paths: &Paths) -> [u8; 32] {
    let path = paths.selfsync_device_file();
    if let Ok(bytes) = std::fs::read(&path) {
        if bytes.len() == 32 {
            let mut id = [0u8; 32];
            id.copy_from_slice(&bytes);
            return id;
        }
    }
    let mut id = [0u8; 32];
    rand::RngCore::fill_bytes(&mut rand::thread_rng(), &mut id);
    let _ = std::fs::write(&path, id);
    id
}

/// Namespaces whose keys are dynamic (set-like) — used to detect LOCAL removals so they
/// propagate as tombstones (unblock, delete contact, leave a circle). Scalar namespaces
/// (`profile:`/`setting:`) are always present, so they're never spuriously removed.
pub const DYNAMIC_PREFIXES: [&str; 3] = ["contact:", "blocked:", "circle:"];

/// The current local state as namespaced key → value bytes (no stamps). `prefs` contributes
/// profile/settings/contacts/blocked; `social` contributes the circle structure (name + member
/// bundles + relay nodes pulled from `prefs.relays`).
pub fn current_local(prefs: &Prefs, social: &HavenSocial) -> BTreeMap<String, Vec<u8>> {
    let mut m: BTreeMap<String, Vec<u8>> = BTreeMap::new();

    // Profile (UTF-8 bytes).
    m.insert("profile:name".into(), prefs.profile.name.clone().into_bytes());
    m.insert("profile:emoji".into(), prefs.profile.emoji.clone().into_bytes());
    m.insert("profile:bio".into(), prefs.profile.bio.clone().into_bytes());
    m.insert("profile:link".into(), prefs.profile.link.clone().into_bytes());

    // Global settings. retention as u64 LE (None = 0 = keep all); host-on-launch as a 1-byte flag.
    let retention = prefs.retention_secs.unwrap_or(0);
    m.insert("setting:retention".into(), retention.to_le_bytes().to_vec());
    m.insert("setting:host_on_launch".into(), vec![if prefs.host_on_launch { 1 } else { 0 }]);

    // Roster: contacts (full card, deterministic serde) + blocked list (marker).
    for c in &prefs.contacts {
        if let Ok(data) = serde_json::to_vec(c) {
            m.insert(format!("contact:{}", c.id_hex), data);
        }
    }
    for hex in &prefs.blocked {
        m.insert(format!("blocked:{hex}"), vec![1]);
    }

    // Explicit circle severances (grow-only) — so a removal converges to our other devices as an
    // INTENTIONAL record rather than being inferred from (strictly-additive) member absence.
    for entry in &prefs.circle_removals {
        m.insert(format!("removal:{entry}"), vec![1]);
    }

    // Circles: name + member bundles + relay nodes, so another device can reconstruct each circle
    // and seal to every member. Encoded via the shared FFI encoder so the bytes are byte-identical
    // to iOS/Android (it base64's the RAW bundles, sorts members/relays, alphabetical-key JSON).
    for ci in social.circles() {
        // Skip DM pseudo-circles — they reconstruct from the contact pair, not from sync.
        if ci.id.starts_with("dm:") {
            continue;
        }
        let member_bundles = social.circle_member_bundles(ci.id.clone());
        let mut relays: Vec<String> = prefs.relays.get(&ci.id).cloned().unwrap_or_default();
        relays.sort();
        relays.dedup();
        let data = encode_circle_sync(ci.name.clone(), member_bundles, relays);
        if !data.is_empty() {
            m.insert(format!("circle:{}", ci.id), data);
        }
    }

    m
}

/// Decode a stored `setting:host_on_launch` (or any 1-byte bool marker) value.
fn bool_value(v: &[u8]) -> Option<bool> {
    v.first().map(|b| *b == 1)
}

/// Apply a converged `AccountState` back into the local stores. Mutates `prefs` + `social`
/// in place, and ONLY where a value actually differs (avoid churn / feedback loops). Returns
/// `true` if anything local changed (so the caller persists + emits `haven:changed`).
///
/// `entries` is the converged state's live `(key, value)` pairs (already collected so the caller
/// can drop the borrow on the CRDT before locking the prefs/social mutexes).
pub fn apply_local(
    entries: &[(String, Vec<u8>)],
    prefs: &mut Prefs,
    social: &HavenSocial,
) -> bool {
    let mut changed = false;

    // Helper closures over `entries`.
    let get = |key: &str| -> Option<&[u8]> {
        entries.iter().find(|(k, _)| k == key).map(|(_, v)| v.as_slice())
    };

    // Profile (UTF-8).
    macro_rules! apply_str {
        ($key:expr, $field:expr) => {
            if let Some(v) = get($key) {
                if let Ok(s) = std::str::from_utf8(v) {
                    if s != $field {
                        $field = s.to_string();
                        changed = true;
                    }
                }
            }
        };
    }
    apply_str!("profile:name", prefs.profile.name);
    apply_str!("profile:emoji", prefs.profile.emoji);
    apply_str!("profile:bio", prefs.profile.bio);
    apply_str!("profile:link", prefs.profile.link);

    // Settings.
    if let Some(v) = get("setting:retention") {
        if v.len() == 8 {
            let mut a = [0u8; 8];
            a.copy_from_slice(v);
            let n = u64::from_le_bytes(a);
            let want = if n == 0 { None } else { Some(n) };
            if want != prefs.retention_secs {
                prefs.retention_secs = want;
                changed = true;
            }
        }
    }
    if let Some(v) = get("setting:host_on_launch") {
        if let Some(b) = bool_value(v) {
            if b != prefs.host_on_launch {
                prefs.host_on_launch = b;
                changed = true;
            }
        }
    }

    // Contacts: upsert everything present; drop locals the converged state no longer has
    // (a contact deleted on another device propagated as a tombstone).
    let mut want_contacts: BTreeMap<String, Contact> = BTreeMap::new();
    for (k, v) in entries {
        if let Some(_id) = k.strip_prefix("contact:") {
            if let Ok(c) = serde_json::from_slice::<Contact>(v) {
                want_contacts.insert(c.id_hex.clone(), c);
            }
        }
    }
    // Upsert.
    for c in want_contacts.values() {
        match prefs.contacts.iter_mut().find(|x| x.id_hex == c.id_hex) {
            Some(existing) => {
                if existing.name != c.name || existing.verify_hex != c.verify_hex {
                    existing.name = c.name.clone();
                    existing.verify_hex = c.verify_hex.clone();
                    changed = true;
                }
            }
            None => {
                prefs.contacts.push(c.clone());
                changed = true;
            }
        }
    }
    // ADDITIVE ONLY — never drop a contact a peer simply doesn't list. Absence-based removal made a
    // freshly-restored (empty) device wipe the primary's contacts/circles/posts (the iOS/Android
    // data-loss bug). Real deletions must propagate as explicit records, not be inferred from absence.

    // Blocked list: reconcile both directions.
    let mut want_blocked: Vec<String> = entries
        .iter()
        .filter_map(|(k, _)| k.strip_prefix("blocked:").map(|h| h.to_string()))
        .collect();
    want_blocked.sort();
    want_blocked.dedup();
    // Add the missing.
    for hex in &want_blocked {
        if !prefs.blocked.contains(hex) {
            prefs.blocked.push(hex.clone());
            changed = true;
        }
    }
    // Remove the extra.
    let before = prefs.blocked.len();
    prefs.blocked.retain(|h| want_blocked.contains(h));
    if prefs.blocked.len() != before {
        changed = true;
    }

    // Explicit circle severances synced from our other devices (grow-only): record them locally so the
    // member isn't re-registered below, and apply the removal to the circle here too.
    for (k, _) in entries {
        let Some(entry) = k.strip_prefix("removal:") else { continue };
        if !prefs.circle_removals.iter().any(|e| e == entry) {
            prefs.circle_removals.push(entry.to_string());
            changed = true;
        }
        if let Some((cid, hex)) = entry.split_once('|') {
            social.remove_from_circle(cid.to_string(), hex.to_string());
        }
    }

    // Circles: reconstruct each synced circle — create it + register every member's bundle so this
    // device can seal to them, and record its relay mailbox(es). STRICTLY ADDITIVE (no absence-based
    // member-prune or circle-leave — that wiped accounts on a freshly-restored device).
    let existing: Vec<(String, String)> =
        social.circles().into_iter().map(|c| (c.id, c.name)).collect();
    for (k, v) in entries {
        let Some(id) = k.strip_prefix("circle:") else { continue };
        let Some(cs) = decode_circle_sync(v.clone()) else { continue };
        match existing.iter().find(|(cid, _)| cid == id) {
            None => {
                social.create_circle(id.to_string(), cs.name.clone());
                changed = true;
            }
            Some((_, cur_name)) => {
                if *cur_name != cs.name {
                    social.rename_circle(id.to_string(), cs.name.clone());
                    changed = true;
                }
            }
        }
        // Register every synced member bundle so this device can seal to them. ADDITIVE — we never
        // remove a member just because a peer's record doesn't list them — but we DO skip anyone we've
        // explicitly severed (anti-reinflation).
        let prefix = format!("{id}|");
        let removed_here: Vec<String> = prefs
            .circle_removals
            .iter()
            .filter_map(|e| e.strip_prefix(&prefix).map(|h| h.to_string()))
            .collect();
        for bundle in &cs.member_bundles {
            let node_hex = p2pcore::identity::HavenId::from_bytes(bundle)
                .ok()
                .map(|hid| hex::encode(hid.node_id_bytes()))
                .unwrap_or_default();
            if !node_hex.is_empty() && removed_here.iter().any(|h| h == &node_hex) {
                continue; // severed — never re-add
            }
            let _ = social.add_contact_bundle(id.to_string(), bundle.clone());
        }
        if !cs.relays.is_empty() {
            let list = prefs.relays.entry(id.to_string()).or_default();
            for node in &cs.relays {
                if !list.contains(node) {
                    list.push(node.clone());
                    changed = true;
                }
            }
        }
    }
    // (No absence-based circle leaving — a circle missing from a peer's slot is NOT a signal to leave
    // it; that destroyed accounts on a freshly-restored device. Explicit leave is intentional only.)
    let _ = &existing;

    changed
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shared_circle_encoder_is_deterministic_and_alphabetical_and_round_trips() {
        // The desktop now defers to the shared FFI encoder for byte-parity with iOS/Android.
        // Raw bundle bytes (the encoder base64's them itself): 0xAA0001 -> "qgAB", 0x000102 -> "AAEC".
        let bundles = vec![vec![0xAAu8, 0x00, 0x01], vec![0x00u8, 0x01, 0x02]];
        let bytes = encode_circle_sync("Home".into(), bundles.clone(), vec!["node1".into()]);
        let json = String::from_utf8(bytes.clone()).unwrap();
        // Alphabetical keys, sorted base64 members ("AAEC" < "qgAB"), matching iOS sortedKeys.
        assert_eq!(json, r#"{"members":["AAEC","qgAB"],"name":"Home","relays":["node1"]}"#);
        // Round-trips back to the same raw bundle bytes (order-independent set membership).
        let rec = decode_circle_sync(bytes).unwrap();
        assert_eq!(rec.name, "Home");
        assert_eq!(rec.relays, vec!["node1".to_string()]);
        assert!(rec.member_bundles.contains(&bundles[0]));
        assert!(rec.member_bundles.contains(&bundles[1]));
    }

    #[test]
    fn dynamic_prefixes_cover_contact_blocked_and_circle() {
        assert!(DYNAMIC_PREFIXES.contains(&"contact:"));
        assert!(DYNAMIC_PREFIXES.contains(&"blocked:"));
        assert!(DYNAMIC_PREFIXES.contains(&"circle:"));
    }
}
